subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -547,6 +516,108 @@ int nrf_cloud_encode_config_response(struct nrf_cloud_data const *const input,
 	return 0;
 }
 
+void nrf_cloud_log_control(int log_level)
+{
+	LOG_DBG("Setting nRF Cloud log level = %d", log_level);
+}
+
+int nrf_cloud_encode_control_response(struct nrf_cloud_data const *const input,
+				      struct nrf_cloud_data *const output,
+				      bool *const has_control)
+{
+	__ASSERT_NO_MSG(output != NULL);
+	__ASSERT_NO_MSG(input != NULL);
+
+	char *buffer = NULL;
+	cJSON *root_obj = NULL;
+	cJSON *desired_obj = NULL;
+	cJSON *reported_obj = NULL;
+	cJSON *state_obj = NULL;
+	cJSON *control_obj = NULL;
+	cJSON *alert_obj = NULL;
+	cJSON *log_obj = NULL;
+	cJSON *input_obj = input ? cJSON_Parse(input->ptr) : NULL;
+
+	if (input_obj == NULL) {
+		return -ESRCH; /* invalid input or no JSON parsed */
+	}
+
+	LOG_DBG("Got delta: %s", (const char *)input->ptr);
+	/* A delta update will have the config inside of state */
+	state_obj = cJSON_DetachItemFromObject(input_obj, JSON_KEY_STATE);
+	control_obj = cJSON_DetachItemFromObject(
+		state_obj ? state_obj : input_obj, JSON_KEY_CTRL);
+	if (control_obj) {
+		LOG_DBG("Shadow delta has control section");
+		alert_obj = cJSON_GetObjectItem(control_obj, JSON_KEY_ALERT);
+		if (cJSON_IsBool(alert_obj)) {
+			nrf_cloud_alert_control(cJSON_IsTrue(alert_obj));
```

Comment:
I modeled this after `nrf_cloud_encode_config_response`, which decodes the shadow delta then reencodes it so it can be sent back to the cloud.

This existing function took a shortcut by not fully parsing every field within the shadow's `config` block, then reencoding them before; instead, it just detached the config_obj (which has many children) from the input JSON and attached it to the output JSON.  Sadly, the user of this section, ATv2, ends up getting a copy of this output JSON, which it then has to parse again to extract the values in this `config` section.

I'm trying to avoid having to parse the JSON twice, or add JSON handling to the alerts part of the nrfcloud library.

Eventually we need to refactor all of this to be able to handle other encodings besides JSON.

Anyway, I'll see if I can figure out a way to do this more cleanly for alerts.

Reply:
I like what you've done to restructure this